File: app/controllers/repositories_controller.rb
require "services/github/connector"

class RepositoriesController < ApplicationController
  def index
  end

  def search
    response = Services::Github::Connector.search_repositories(query_params)
    if response[:status_code] == 200
      repositories = response["items"]
      total_count = response["total_count"]
      page = 1
      render json: { repositories: repositories, fetched_count: repositories.count, total: total_count, page: page }
    else
      render json: { error: response[:message] }, status: response[:status_code]
    end
  end

  private
  def query_params
    params.permit(:query).fetch(:query, nil)
  end
end

File: app/javascript/controllers/repositories_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = [ "query", "results", "total" ]

  search(event) {
    event.preventDefault()
    // remove existing results
    this.resultsTarget.innerHTML = "";
    this.totalTarget.innerHTML = "";
    // fetch new results from RepositoryController#search
    fetch(`/repositories/search?query=${this.queryTarget.value}`)
      .then((response) => response.json())
      .then(data => {
        if (data['error'] != null) {
          let li = document.createElement("li");
          li.innerHTML = data['error'];
          this.resultsTarget.appendChild(li);
          return;
        }
        let total = data['total'];
        let fetched_count = data['fetched_count'];
        this.totalTarget.innerHTML = `Showing ${fetched_count} of total ${total}:`;

        data['repositories'].forEach(element => {
          let li = document.createElement("li");
          let a = document.createElement("a");
          a.setAttribute("href", element['url']);
          a.innerHTML = element['name'];
          a.setAttribute("target", "_blank");
          li.appendChild(a);
          this.resultsTarget.appendChild(li);
        })
      })
  }
}
